cartographer_ros/cartographer_ros/assets_writer.cc

```diff
@@ -0,0 +1,241 @@
+#include "cartographer/io/proto_stream.h"
+#include "cartographer/mapping/proto/pose_graph.pb.h"
+#include "cartographer/mapping/proto/trajectory_builder_options.pb.h"
+#include "cartographer/sensor/point_cloud.h"
+#include "cartographer/sensor/range_data.h"
+#include "cartographer/transform/transform_interpolation_buffer.h"
+#include "cartographer_ros/msg_conversion.h"
+#include "cartographer_ros/ros_map_writing_points_processor.h"
+#include "cartographer_ros/split_string.h"
+#include "cartographer_ros/time_conversion.h"
+#include "cartographer_ros/urdf_reader.h"
+#include "gflags/gflags.h"
+#include "glog/logging.h"
+#include "ros/ros.h"
+#include "ros/time.h"
+#include "rosbag/bag.h"
+#include "rosbag/view.h"
+#include "tf2_eigen/tf2_eigen.h"
+#include "tf2_msgs/TFMessage.h"
+#include "tf2_ros/buffer.h"
+#include "urdf/model.h"
+
+namespace cartographer_ros {
+
+constexpr char kTfStaticTopic[] = "/tf_static";
+namespace carto = ::cartographer;
+
+template <typename T>
+std::unique_ptr<carto::io::PointsBatch> HandleMessage(
+    const T &message, const std::string &tracking_frame,
+    const tf2_ros::Buffer &tf_buffer,
+    const carto::transform::TransformInterpolationBuffer
+        &transform_interpolation_buffer) {
+  const carto::common::Time start_time = FromRos(message.header.stamp);
+
+  auto points_batch = carto::common::make_unique<carto::io::PointsBatch>();
+  points_batch->start_time = start_time;
+  points_batch->frame_id = message.header.frame_id;
+
+  ::cartographer::sensor::PointCloudWithIntensities point_cloud;
+  ::cartographer::common::Time point_cloud_time;
+  std::tie(point_cloud, point_cloud_time) =
+      ToPointCloudWithIntensities(message);
+  CHECK_EQ(point_cloud.intensities.size(), point_cloud.points.size());
+
+  for (size_t i = 0; i < point_cloud.points.size(); ++i) {
+    const carto::common::Time time =
+        point_cloud_time + carto::common::FromSeconds(point_cloud.points[i][3]);
+    if (!transform_interpolation_buffer.Has(time)) {
+      continue;
+    }
+    const carto::transform::Rigid3d tracking_to_map =
+        transform_interpolation_buffer.Lookup(time);
+    const carto::transform::Rigid3d sensor_to_tracking =
+        ToRigid3d(tf_buffer.lookupTransform(
+            tracking_frame, message.header.frame_id, ToRos(time)));
+    const carto::transform::Rigid3f sensor_to_map =
+        (tracking_to_map * sensor_to_tracking).cast<float>();
+    points_batch->points.push_back(sensor_to_map *
+                                   point_cloud.points[i].head<3>());
+    points_batch->intensities.push_back(point_cloud.intensities[i]);
+    // We use the last transform for the origin, which is approximately correct.
+    points_batch->origin = sensor_to_map * Eigen::Vector3f::Zero();
+  }
+  if (points_batch->points.empty()) {
+    return nullptr;
+  }
+  return points_batch;
+}
+
+void RunAssetsWriterPipeline(const std::string &pose_graph_filename,
```

Comment:
Revert the change from `Foo& foo` to `Foo &foo` here and everywhere else in the file. We should run clang-format with `--style=Google`. Maybe we can add a `.clang-format` file to make this the default for this repository. This probably should be done in a separate PR though.